oak_restricted_kernel/src/syscall/fd.rs

```diff
@@ -0,0 +1,107 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use alloc::{
+    boxed::Box,
+    collections::{btree_map::Entry, BTreeMap},
+    slice,
+};
+use core::ffi::{c_int, c_size_t, c_ssize_t, c_void};
+use oak_core::sync::OnceCell;
+use oak_restricted_kernel_interface::Errno;
+use spinning_top::Spinlock;
+
+pub trait FileDescriptor: Send {
+    fn read(&mut self, buf: &mut [u8]) -> Result<isize, Errno>;
+    fn write(&mut self, buf: &[u8]) -> Result<isize, Errno>;
+    fn sync(&mut self) -> Result<(), Errno>;
+}
+
+type Fd = c_int;
+
+static FILE_DESCRIPTORS: OnceCell<Spinlock<BTreeMap<Fd, Box<dyn FileDescriptor>>>> =
```

Comment:
nit: I don't think you need a OnceCell here. Spinlock has a `const_spinlock` function and `BTreeMMap::new` is also const. It will remove the need to all the lazy initialisation below.

Reply:
Ah, cool, thanks, I missed those!